### ISSUE TITLE
augment() generic should have arguments x and data

### ISSUE THREAD

**Issue by hadley** [dgrtwo/broom#137]
If `augment()` is truly about augmenting data, then that should be explicit in the function signature. In general, models will not always contain the data used to fit them (especially when that data is large or lives outside of R), and it's better to be explicit.

I think the design of `resid()` in R encourages slightly sloppy thinking about residuals - it makes you think that residuals are a property of a model, when actually they are a product of the model and a dataset.

**Comment by dgrtwo** [comment]
I originally had `data` in the augment signature, but [removed it with this commit](c974d2d3b9672dcb589154ffaf236c9259f2d80b) in order to make `rowwise_df_tidiers` work, where rather than data the second argument needed to be a bare column name, with `data` a possible third argument.

I definitely understand that do + rowwise df tidiers are no longer the way to fit multiple models (rather the nest/unnest/purrr patterns) but the shortcut is used frequently enough that I think there's a high cost to deprecate.

I do agree philosophically that every `augment` method besides special ones should have `data` argument second.

**Comment by hadley** [comment]
Yeah, I agree that it will be painful to deprecate but it's worth it in the long run.

**Comment by alexpghayes** [comment]
It seems like a good time to revisit this: I'm also in favor of deprecating rowwise tidiers (#154) and adding in a `data` argument to `augment`.

Some of the existing augment methods have both `data` and `newdata` arguments, which I think it confusing and probably not necessary (for example `augment.prcomp`, #317). Settling on just one would be a good idea. I'm a fan of `newdata` because of the allusion to `predict(model, newdata)`.
